From 8f931cbf36ce6d356b771a74fcd2fd2e48d599d7 Mon Sep 17 00:00:00 2001
From: Vidar Tonaas Fauske <vidartf@gmail.com>
Date: Wed, 29 Sep 2021 11:02:04 +0100
Subject: [PATCH 1/2] Replace innerHTML with innerText

For any content where the user can potentially influence the content.
---
 packages/labextension/src/widget.ts | 6 ++++--
 packages/nbdime/src/common/util.ts  | 2 +-
 packages/webapp/src/app/diff.ts     | 7 +++++--
 packages/webapp/src/app/merge.ts    | 5 ++++-
 4 files changed, 14 insertions(+), 6 deletions(-)

diff --git a/packages/labextension/src/widget.ts b/packages/labextension/src/widget.ts
index c1055d5b..7be02006 100644
--- a/packages/labextension/src/widget.ts
+++ b/packages/labextension/src/widget.ts
@@ -239,9 +239,11 @@ namespace Private {
         <button class="nbdime-export" style="display: none">Export diff</button>
       </div>
       <div class=nbdime-header-banner>
-        <span class="nbdime-header-base">${baseLabel}</span>
-        <span class="nbdime-header-remote">${remoteLabel}</span>
+        <span class="nbdime-header-base"></span>
+        <span class="nbdime-header-remote"></span>
       </div>`;
+    (node.getElementsByClassName("nbdime-header-base")[0] as HTMLSpanElement).innerText = baseLabel;
+    (node.getElementsByClassName("nbdime-header-remote")[0] as HTMLSpanElement).innerText = remoteLabel;
 
     return new Widget({node});
   }
diff --git a/packages/nbdime/src/common/util.ts b/packages/nbdime/src/common/util.ts
index 9d118c7e..7704246a 100644
--- a/packages/nbdime/src/common/util.ts
+++ b/packages/nbdime/src/common/util.ts
@@ -298,7 +298,7 @@ function buildSelect(options: string[], select?: HTMLSelectElement): HTMLSelectE
   }
   for (let option of options) {
     let opt = document.createElement('option');
-    opt.value = opt.innerHTML = option;
+    opt.value = opt.innerText = option;
     select.appendChild(opt);
   }
   return select;
diff --git a/packages/webapp/src/app/diff.ts b/packages/webapp/src/app/diff.ts
index edc6a2dc..b32c672d 100644
--- a/packages/webapp/src/app/diff.ts
+++ b/packages/webapp/src/app/diff.ts
@@ -180,11 +180,14 @@ function onDiffRequestCompleted(data: any) {
  */
 function onDiffRequestFailed(response: string) {
   console.log('Diff request failed.');
-  let root = document.getElementById('nbdime-root');
+  const root = document.getElementById('nbdime-root');
   if (!root) {
     throw new Error('Missing root element "nbidme-root"');
   }
-  root.innerHTML = '<pre>' + response + '</pre>';
+  const pre = document.createElement('pre');
+  pre.innerText = response;
+  root.innerHTML = '';
+  root.appendChild(pre);
   diffWidget = null;
   toggleSpinner(false);
 }
diff --git a/packages/webapp/src/app/merge.ts b/packages/webapp/src/app/merge.ts
index 22941ab9..00856c19 100644
--- a/packages/webapp/src/app/merge.ts
+++ b/packages/webapp/src/app/merge.ts
@@ -179,7 +179,10 @@ function onMergeRequestFailed(response: string) {
   if (!root) {
     throw new Error('Missing root element "nbidme-root"');
   }
-  root.innerHTML = '<pre>' + response + '</pre>';
+  const pre = document.createElement('pre');
+  pre.innerText = response;
+  root.innerHTML = '';
+  root.appendChild(pre);
   mergeWidget = null;
   toggleSpinner(false);
 }

From 19fded3bb9ac3405d1efa4d34b9b45d3e261dcc0 Mon Sep 17 00:00:00 2001
From: Vidar Tonaas Fauske <vidartf@gmail.com>
Date: Thu, 30 Sep 2021 12:22:20 +0100
Subject: [PATCH 2/2] Fix buildSelect + tests

---
 packages/nbdime/src/common/util.ts           |  2 +-
 packages/nbdime/test/src/common/util.spec.ts | 30 ++++++++++++++++++++
 2 files changed, 31 insertions(+), 1 deletion(-)

diff --git a/packages/nbdime/src/common/util.ts b/packages/nbdime/src/common/util.ts
index 7704246a..e4689218 100644
--- a/packages/nbdime/src/common/util.ts
+++ b/packages/nbdime/src/common/util.ts
@@ -298,7 +298,7 @@ function buildSelect(options: string[], select?: HTMLSelectElement): HTMLSelectE
   }
   for (let option of options) {
     let opt = document.createElement('option');
-    opt.value = opt.innerText = option;
+    opt.text = option;
     select.appendChild(opt);
   }
   return select;
diff --git a/packages/nbdime/test/src/common/util.spec.ts b/packages/nbdime/test/src/common/util.spec.ts
index 46fb2fa6..66715da8 100644
--- a/packages/nbdime/test/src/common/util.spec.ts
+++ b/packages/nbdime/test/src/common/util.spec.ts
@@ -255,6 +255,36 @@ describe('common', () => {
 
     });
 
+    describe('buildSelect', () => {
+
+      it('should create an empty select', () => {
+        let value = util.buildSelect([]);
+        expect(value.outerHTML).toEqual("<select></select>");
+      });
+
+      it('should reuse a given select', () => {
+        const select = document.createElement('select');
+        let value = util.buildSelect([], select);
+        expect(value).toBe(select);
+      });
+
+      it('should create a select with options', () => {
+        let value = util.buildSelect([
+          'foo',
+          'bar',
+          '<div>boo</div>'
+        ]);
+        expect(value.outerHTML).toEqual(
+          '<select>' +
+            '<option>foo</option>' +
+            '<option>bar</option>' +
+            '<option>&lt;div&gt;boo&lt;/div&gt;</option>' +
+          '</select>'
+        );
+      });
+
+    });
+
   });
 
 });
